howto reorg

JFRudzinski opened the pull request:
**Problem:** How-to Guides are difficult for users to navigate through. Context of usage is unclear.

- Overview pages categorizes Users vs. All Other Roles, with user customization being lumped into the latter

- No clear distinction between Customization (mostly YAML-based customization) and Plugins (Python-based customization) -- also some mis-categorized how-tos in this context

**Solution:** Restructure How-to Guides to clarify groupings / usage role:
    
- deepen nav bar tree by 1 with overarching categories for the 3 usage roles (note that previously we explicitly avoided over-nesting, so this is worth carefully assessing)
    
- add icons to indicate usage roles (optional + note that what I implemented is simply to demonstrate the idea, if we want icons we need to design nicer looking ones :wink: )
   
- update how-to overview page structure to provide more context for usage/relevance

![howto_restructure](https://github.com/user-attachments/assets/ec4733dd-df9c-478e-b2f8-cc3218631e87)

@siamakn FYI

lukaspie commented on docs/howto/overview.md:
```diff
@@ -24,43 +26,47 @@ Use NOMAD to manage, explore, and analyze data.
+### YAML (quick)
 
-Self-hosting NOMAD for your lab or institution.
+Implement quick and dirty customizationations of NOMAD entries.
```
Maybe would be good to add here when you want to do quick customizationations and when it makes sense to go for a plugin.

ahm531 commented:
Thanks Joe for preparing this. Here are my thoughts:

1. I recommend adding an introductory paragraph to the overview page. Since many readers are unfamiliar with the Diátaxis framework, it’s important to briefly explain the purpose of each section. We already explain this on the `index.md` page, but including it on each overview page would be valuable. We have already done this for the tutorials overview page, /tutorial/overview.md.
I also suggest planning to create overview pages for the Explanation and Reference sections. What do you think?

2. The structure you created for the how-to guides is effective and aligns with our recent approach to organizing tutorials and training materials. To maintain consistency, I suggest using similar terms and categorizing topics according to their usage categories. This would provide a guiding framework across the documentation.
For example, on the Tutorials overview page `/tutorial/overview.md`, NOMAD usage is categorized as *Basic Usage*, *Customization*, and *Self-Hosting*. The sections in the How-to-Guides overview resemble these categories. Should we unify the terms we use to describe the usage categories? This also inculde the possibility of renaming the catogerize on the tutorials page. 
In an early draft, the tutorials had two customization levels. After discussing this with @lauri-codes  and @hampusnasstrom, we decided that customization with YAML and tabular parsers to be part of the *Basic Usage* category (however it wasn't mentioned explicitly in the descriptions) while the `Customization` category only covers NOMAD plugins. I think we should keep this approach in the how-to pages as well. Since YAML files (ELN schemas or tabular parsers) are typically uploaded via the GUI (let me know if this is not entirely correct), they could also fit into the *Manage and Find Data/GUI* section.

4. I agree with the above comment about the tags. They can be confusing, especially if readers jump into the pages without first reading the overview in the How-to-Guides. At the same time, I think explaining the roles is useful, as you did in the sentences under the main usage categories. Perhaps we can discuss how to clarify the relationship between roles and usage categories on the `index.md` page of the documentation.
Personally, I’m hesitant to assign roles to specific topics in our documentation since our successful NOMAD use cases are achieved by one person who wears many hats. I suggest keeping the focus on usage categories while clarifying how they relate to roles. This way, all the pages in the documentation remain welcoming to everyone.

5. How about separating Hosting from Developing NOMAD? That would give us 4 sections. I believe development is mostly done by our team and a few motivated external users, so it doesn't represent a large part of the user base. Additionally, I don't see much connection between self-hosting and NOMAD development. Keeping them separate would clarify the structure.

lauri-codes commented:
Something I realized after @ndaelman-hu pointed out that we don't really have installation instructions: I think an immediate reaction from a new user would indeed be to find the how-to-guide on "how can I install NOMAD on a computer". Currently we don't have a separate section for installation, which makes it hard for people to find this information: basically it can be found under "NOMAD Oasis", but this terminology is not clear at all for a new user.

I also agree with @ahm531 that we should not mix "hosting and developing" into the same section: these are two very different things.

I would thus propose the following structure

- Overview
- Manage and find data
- Install (how to configure and run a NOMAD Oasis, hosting, How to install our Python package)
- Customize (Plugins, YAML)
- Develop (development guidelines that are common for plugins and `nomad-FAIR`)

Thoughts?

ndaelman-hu commented:
> I have been looking through this for a while now. It is a bit difficult for me to reconcile some of the suggestions. I try to document my main thoughts below, but I think we need a meeting to discuss this again now.
> 
> **Installation** I'm not sure if I am missing something. We do have simple [installation instructions](https://fairmat-nfdi.github.io/nomad-docs/howto/programmatic/pythonlib.html), it is just a bit hidden in the Programmatic usage section. Is there additional content associated with install nomad-lab? If not, I believe it fits nicely at the top of the Programmatic Usage section. It is already referenced within the first how-to of this section.
> 
> Alternatively, as Laurie suggested, we could have something like:
> 
> ```yaml
>       - Install:
>           - nomad-lab: howto/programmatic/pythonlib.md
>           - NOMAD Oasis:
>             - Configure an Oasis: howto/oasis/configure.md
>             - Deploy an Oasis: howto/oasis/deploy.md
>             - Update an Oasis: howto/oasis/update.md
>             - Perform admin tasks: howto/oasis/admin.md
> ```
> 
> I don't love that this obscures a bit the various parts of the Oasis subsection, which is beyond installation. I feel that "NOMAD Oasis" should be a stand-alone section under the root of How-to.
> 
> We could instead have a standalone single page "Install nomad-lab", but the main question for me is for what reason do we expect users to install nomad-lab. If we make Installation more prominent it may indicate to people that they need to install for basic usage, so we need a good balance in that sence.
> 
> **Customization with YAML** I don't really know what to do with this. I agree with Laurie that if under customization it should be placed second to encourage proper customization. However, something like the "define workflows" page is, at least at the moment, encouraged usage. I also agree with Ahmed that in principle it could go under GUI usage, however, in practice I don't really like this because again it sort of implies that it is a suggested usage.
> 
> Perhaps certain pages, e.g., "define workflows" we can move to GUI usage, while others about defining YAML schemas we keep as a secondary customization section?
> 
> **Develop Section** It is very difficult for me to assess since these pages are somewhat outdated or redundant with other content...or perhaps I just don't know the content well enough. However, I do not think we should mix content about specifically contributing to nomad-lab with best practices for general dev (i.e., plugins). This might mean that we need something like a "Development Guidelines" section.
> 
> **Simpler assignment discrepancies/disagreements**
> 
>     * "work with units", "use hdf5", "use mapping annotations" --- where do they live if not under plugins?

Regarding missing "Installation" instructions, I (and the user) were referring to the docker deployments, both for an Oasis and the developer distro. The link you provide only covers the Python module. It is also a bit too hidden: most software have a clearly denoted chapter.

lauri-codes commented:
I guess the main problem is that just installing the Python library will not get you anywhere, it is just a part of a NOMAD installation. Thus this should not be our main install instruction.

We could try to combine both having a clear installation section and an Oasis section with the following:

```yaml
  - Install and manage a NOMAD Oasis:
    - Install: howto/install/overview.md  # Would explain what a NOMAD installation means, contains links to `nomad-distro-template`+`nomad-distro-dev`, also contains a small section on the Python library
    - Configure: howto/oasis/configure.md
    - Deploy: howto/oasis/deploy.md
    - Update: howto/oasis/update.md
    - Administator tasks: howto/oasis/admin.md
```

But I agree that we could discuss this in a small group.

ahm531 commented:
Thanks a lot for summarizing the discussion into the drafted structure. Here are some points of potential improvement:

**Manage and find data**

- The two subsections could be phrased as How-to questions (e.g. Using the GUI and Using Code). The third level could remain topical statements or also be phrased as How-to questions. 
- Use ELNs → Document experiments using ELNs?
- Use NORTH → Analyze data using NORTH?
- Define ELNs → Create a custom ELN?
- For the Graph-Style API (and its nested topics), I suggest starting with verbs (when possible).
- Would it make sense to reorder the list so that related topics are adjacent? For example: Upload and Publish then Parse tabular data; Use ELNs then Create a custom ELN.

**Develop Python Plugins**

- Do we need Python in the title? It might be confusing — NOMAD Plugins vs. Python Plugins (plugins written in Python or plugins specifically for Python?). I would suggest Developing Plugins.

- Instead of Types, would it be more precise to call them Entry Points? I believe this is the term we use officially, right? I assume that users looking to develop a plugin would not be intimidated by it. What do you think?

- The list of plugin entry points is alphabetical. Conceptually, though, the flow might be clearer as: Schema → Parsers → Apps → "Other entry points". That way, it follows the natural order of defining structure, reading files, and then visualizing/exploring data.

**Install and manage a NOMAD Oasis**

- Hampus suggested earlier the phrase "Hosting a NOMAD Oasis". Perhaps we could use this wording to cover both installation and management?

- Admin tasks could be phrased as Administer an Oasis or Manage an Oasis.

JFRudzinski commented:
Sorry for the delay on this... Here is my new working version with changes based on your comments:

```yaml
  - How-to guides:
      - Overview: howto/overview.md # TODO update once structure is finalized
      - Manage and find data:
          - With the GUI:
              - Upload and publish data: howto/manage/upload.md
              - Enter data with ELNs: howto/manage/eln.md
              - Explore data: howto/manage/explore.md
              - Analyze data in NORTH: howto/manage/north.md # TODO rename depending on how-to content
              - Create custom workflows: howto/customization/workflows.md
              - Write a YAML schema package: howto/customization/basics.md # TODO consider consolidating with parts of the next 2 pages
              - Create custom ELNs: howto/customization/elns.md
              - Parse tabular data: howto/customization/tabular.md
          - Programmatically:
              # - Install nomad-lab: howto/programmatic/pythonlib.md # TODO Integrate into "Use the API"
              - Overview: howto/programmatic/api.md # TODO separate into How-to and Explanation/Reference
              - Download data: howto/programmatic/download.md
              - Publish data using Python: howto/programmatic/publish_python.md
              - Access processed data: howto/programmatic/archive_query.md
              - Transform data: howto/programmatic/json_transformer.md
              - Make Graph-style API Calls: # TODO consolidate into a single page with supporting explanation/reference material
                  - Basics: howto/graph-api/basics.md
                  - Navigation and fuzzy query/fetch: howto/graph-api/graph.structure.md
                  - Handling archive data: howto/graph-api/query.archive.md
                  - Other convenience functionalities: howto/graph-api/misc.md
      - Develop plugins:
          - Overview: howto/plugins/plugins.md # TODO split former Introduction to plugins into short guiding overview + explanation page
          # - Set up a development environment: howto/plugins/setup.md # TODO distro-dev setup + best practices
          - Plugin Types:
              - Actions: howto/plugins/actions.md
              - APIs: howto/plugins/apis.md
              - Apps: howto/plugins/apps.md
              - Example uploads: howto/plugins/example_uploads.md
              - Normalizers: howto/plugins/normalizers.md
              - Parsers: howto/plugins/parsers.md
              - Schemas: howto/plugins/schema_packages.md
          - Built-In Tools:
              - Work with units: howto/customization/units.md
              - Handle large data: howto/customization/hdf5.md
              - Parse efficiently: howto/customization/mapping_parser.md # TODO Add clearly scope at the top of this page
      - Host a NOMAD Oasis:
          # - Install: howto/install/install.md # TODO Would explain what a NOMAD installation means, contains links to `nomad-distro-template`+`nomad-distro-dev`, also contains a small section on the Python library
          - Configure an Oasis: howto/oasis/configure.md
          - Deploy an Oasis: howto/oasis/deploy.md
          - Update an Oasis: howto/oasis/update.md
          - Perform Admin tasks: howto/oasis/admin.md
      - Develop the core software: # TODO strip down to only nomad-lab-specific development
          - Get started: howto/develop/setup.md
          - Navigate the code: howto/develop/code.md
          - Contribute: howto/develop/contrib.md
          - Write a normalizer: howto/develop/normalizing.md
          - Extend the search: howto/develop/search.md
          - Make a release: howto/develop/release.md
 ```

And here are detailed responses to each comment:

From @lauri-codes:

> In the programmatic use we should not call the first page "Install nomad-lab", as this gives the impression that this is the only way for programmatic use. Instead we should make it about setting up the different options for using our API (requests, curl, ArchiveQuery), which happens to include also installing nomad-lab.

Done. Commented out Install page with added TODO. The new first page renamed to "Overview". I am 100% on board with this approach, just to mention that this does not help the raised issue about findability of installation instructions (personally I do not think this is an issue)

> In many cases I don't think we can make a clean split between "schema tools" and "parser tools": e.g. units and HDF5 are important for both. I would put all of these under the same subsection, e.g. "Common tools".

Agreed. I named this "Built-In Tools", but we can rename.

From @ahm531:

First a general comment on the How-to phrasing: I agree that we should be consistent with the verb tense (when using a verb), always in "simple tense" not "-ing". Then the title of the page becomes "How-to" + "<navigation label>" + "<can include additional descriptive words". However, I do not think that we should be too strict about forcing the How-to phrasing at every level. At the categories level, e.g., GUI, Programmatic, Plugin Types, and below in some cases it is unnecessary and somewhat awkward to carry through the How-to phrasing. We also already allow "Overview" for introductory sections of some categories. 

In the end, as long as the meaning is clear to the user, and the page title contains the How-to phrasing, I think we should allow the simpler noun-only navigation labels. That being said, I did try to adjust many based on your comments.

Also, in general I think we should keep the nav bar as concise as possible while keeping some contextual clarity, I removed some (IMO) unnecessary words with this in mind. 

> **Manage and find data**
> ...

- See my renaming 
    - I do not want to address the sub-levels of the graph API as these will go away
    - I do not want to imply that ELNs are only for experiments
- I have ordered these in terms of complexity in a way, with the last 3 dealing essentially with creating a custom ELN, but step by step. For me this makes the most sense. In particular for "Parse Tabular Data": you need to define a schema to do this, so it MUST follow "Write a YAML schema package". Also Note that I have a TODO about possibly consolidating a couple or all three of these pages.

> **Develop Python Plugins**
> ...

- Removed "Python"
- I think even if you say Entry Points you need to say Entry Point Types ... @lauri-codes ? I don't have a super strong opinion, only that saying entry points is bringing something technical there which I don't think we need to include in the nav bar.
- I would be fine with reordering the plugin type if people think it will help users understand the process .. @lauri-codes ?

>  **Install and manage a NOMAD Oasis**

- See renaming

lauri-codes commented:
- Unfortunately we can't say "Plugin Types": NOMAD plugin is the Python package that can be installed, and  ["entry point" is the official name](https://packaging.python.org/en/latest/specifications/entry-points/) for the mechanism through which a Python package can expose one or multiple discoverable pieces of customization. So I feel we need to use the technical term "Plugin entry point types" in order to be clear.
- We once had them ordered in a more logical sense, but then in the docs hackathon we got a comment that the ordering seemed arbitrary and we made it alphabetical :man_shrugging: I think everyone has an opinion on what are the most importants entry point types, so I would rather like to keep it alphabetical and thus neutral.

JFRudzinski commented:
Updated:

```yaml
  - How-to guides:
      - Overview: howto/overview.md # TODO update once structure is finalized
      - Manage and find data:
          - With the GUI:
              - Upload and publish data: howto/manage/upload.md
              - Enter data with ELNs: howto/manage/eln.md
              - Explore data: howto/manage/explore.md
              - Analyze data in NORTH: howto/manage/north.md # TODO rename depending on how-to content
              - Create custom workflows: howto/customization/workflows.md
              - Write a YAML schema package: howto/customization/basics.md # TODO consider consolidating with parts of the next 2 pages
              - Create custom ELNs: howto/customization/elns.md
              - Parse tabular data: howto/customization/tabular.md
          - Programmatically:
              # - Install nomad-lab: howto/programmatic/pythonlib.md # TODO Integrate into "Use the API"
              - Overview: howto/programmatic/api.md # TODO separate into How-to and Explanation/Reference
              - Download data: howto/programmatic/download.md
              - Publish data using Python: howto/programmatic/publish_python.md
              - Access processed data: howto/programmatic/archive_query.md
              - Transform data: howto/programmatic/json_transformer.md
              - Make Graph-style API Calls: # TODO consolidate into a single page with supporting explanation/reference material
                  - Basics: howto/graph-api/basics.md
                  - Navigation and fuzzy query/fetch: howto/graph-api/graph.structure.md
                  - Handling archive data: howto/graph-api/query.archive.md
                  - Other convenience functionalities: howto/graph-api/misc.md
      - Develop plugins:
          - Overview: howto/plugins/plugins.md # TODO split former Introduction to plugins into short guiding overview + explanation page
          # - Set up a development environment: howto/plugins/setup.md # TODO distro-dev setup + best practices
          - Entry point types:
              - Actions: howto/plugins/actions.md
              - APIs: howto/plugins/apis.md
              - Apps: howto/plugins/apps.md
              - Example uploads: howto/plugins/example_uploads.md
              - Normalizers: howto/plugins/normalizers.md
              - Parsers: howto/plugins/parsers.md
              - Schemas: howto/plugins/schema_packages.md
          - Built-in tools:
              - Work with units: howto/customization/units.md
              - Handle large data: howto/customization/hdf5.md
              - Parse efficiently: howto/customization/mapping_parser.md # TODO Add clearly scope at the top of this page
      - Host a NOMAD Oasis:
          # - Install: howto/install/install.md # TODO Would explain what a NOMAD installation means, contains links to `nomad-distro-template`+`nomad-distro-dev`, also contains a small section on the Python library
          - Configure an Oasis: howto/oasis/configure.md
          - Deploy an Oasis: howto/oasis/deploy.md
          - Update an Oasis: howto/oasis/update.md
          - Perform Admin tasks: howto/oasis/admin.md
      - Develop the core software: # TODO strip down to only nomad-lab-specific development
          - Get started: howto/develop/setup.md
          - Navigate the code: howto/develop/code.md
          - Contribute: howto/develop/contrib.md
          - Write a normalizer: howto/develop/normalizing.md
          - Extend the search: howto/develop/search.md
          - Make a release: howto/develop/release.md
```

It seems like we are more or less on the same page in terms of the overall structure. If you have any further comments/suggestions, please add them here. In the mean time, I will start reorganizing the actual file locations and references to match the new structure

JFRudzinski commented:
Restructuring is complete! This PR is ready for a full review. @ahm531 perhaps you can look in detail and anyone else can provide supplementary comments?

In the mean time, I will work on adding/adjusting issues based on the topics that arose in the PR and work towards prioritizing/assigning them.
